### ISSUE TITLE
performance leak in self.call()

### ISSUE THREAD

**Issue by spaceone** [circuits/circuits#86]
time python2 slow.py
real    0m6.487s
user    0m0.360s
sys     0m0.030s

```
from circuits import *

class a(Event): pass
class b(Event): pass

class Slow(BaseComponent):
	@handler('started')
	def started(self, manager):
		self.iter = self.foo()
		self.fire(a())

	def foo(self):
		for d in range(30):
			print d
			yield d

	@handler('a')
	def _on_a(self):
		try:
			d = next(self.iter)
		except StopIteration:
			self.stop()
			return
		yield self.call(b(d))
		yield self.call(a())

	@handler('b')
	def _on_b(self, d):
		print d

Slow().run()
```

If you replace "yield self.call" by self.fire you will get:
real    0m0.256s

https://github.com/spaceone/circuits.http/commit/ef5d7f6c02af9cb541982f892848a005ae700c28

**Comment by prologic** [comment]
:+1: Worth looking in to :)

I suspect it *might* have something to do with the fact that coroutines are handled in the "next tick" so to speak; whereas fired events are "queued up" for immediately handling.

But maybe we *can* improve this?

**Comment by prologic** [comment]
With a much simpler implementation of this "slow test":

```
#!python
#!/usr/bin/env python


from circuits import Component, Event


class counter(Event):
    """counter Event"""

    complete = True


class Slow(Component):

    def started(self, *args):
        #  self.fire(counter(10))
        yield self.call(counter(10))

    def counter(self, n):
        for i in xrange(n):
            print i
            yield i

    def counter_complete(self, *args):
        raise SystemExit(0)

Slow().run()
```

I only see a ~10% performance drop in using ``self.call()`` which IHMO is acceptable seeing what it's doing and all :) Thoughts?

**Comment by spaceone** [comment]
That code does not what my code does. You are call()-ing 1 time not 10/30 times (which is a valid use case). Where then 10% is very much. At least it must be documented, also using call() is useless then if it is in the endeffect slower than just fire()ing.

**Comment by prologic** [comment]
Hmm I'll have to study your *contrived example* a bit more :) I'm not quite following it yet :)

**Comment by spaceone** [comment]
Real life code is always better: spaceone/circuits.http@ef5d7f6

**Comment by prologic** [comment]
Yeah okay I see :) I'll have to play with this some more and see what's going on. To be honest I'm not sure why you're using coroutines (*or were trying to*) there anyway?

**Comment by spaceone** [comment]
Well, I thought 'call' is not just equivalent to 'yield self.wait(self.fire(event).event)'. I thought it would just call the event handlers synchroniously. Now I looked at the code and I am smarter ;)

**Comment by prologic** [comment]
Ahh so in your particular case you were calling the wrong API method? ``yield self.wait()`` is indeed **NOT** equivalent to ``self.call()`` :) Both in effect create coroutines!

**Comment by spaceone** [comment]
What are the exact differences between `call()` and `wait()`? In my particular case I want to call the handlers as soon as possible. I probably misunderstood how `call()` is meant to be used?

**Comment by prologic** [comment]
Yeah so ``call/wait`` are essentially the same. ``.call()`` does a ``.fire()`` and ``.wait()``, whilst ``.wait()`` only wiats for an event's completion. Both are used to write synchronous code that is otherwise asynchronous underneath, both turn the event handler into a coroutine (*because you **have** to ``yield``*).

**Comment by prologic** [comment]
Any further thoughts/comments on this? Do we need to change/fix/update anything here? Improve the documentation around this perhaps?

**Comment by prologic** [comment]
@spaceone Do you have any thoughts on improving this at all -- Or should we just document it better? ``call/wait`` provide coroutines (*cooperative threading*); I would never expect them to be as performant as "fire and forget". I have however observed the same behaviour as what you've described in that using coroutiens does cost a bit of performance for the sake of "synchronization".

**Comment by spaceone** [comment]
Maybe we could implement a method `self.fastcall()` ?

**Comment by prologic** [comment]
> Maybe we could implement a method self.fastcall() ?

I'm not sure; but it would be backwards compatible at least. New API.
